File: Hashtable/tasks/threeSum.js
/**
 * 用JavaScript 实现三数之和
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
  if (nums.length < 3) {
    return [];
  }
  let res = [];
  // 数组排序
  nums.sort(function(a, b) {
    return a - b;
  });
  console.log(nums);

  for (let i = 0; i < nums.length - 2; i++) {
    if (i > 0 && nums[i] == nums[i - 1]) {
      continue;
    }

    let left = i + 1;
    let right = nums.length - 1;

    while (left < right) {
      let sum = nums[i] + nums[left] + nums[right];

      if (sum === 0) {
        res.push([nums[i], nums[left], nums[right]]);

        while (left < right && nums[left] === nums[left + 1]) {
          left = left + 1;
        }
        while (left < right && nums[right] === nums[right - 1]) {
          right = right - 1;
        }
        (left += 1), (right -= 1);
      } else if (sum > 0) {
        right--;
      } else if (sum < 0) {
        left++;
      }
    }
  }
  return [...res];
};

console.log(threeSum([0, 0, 0, 0]));
console.log(threeSum([-4, -2, -2, -2, 0, 1, 2, 2, 2, 3, 3, 4, 4, 6, 6]));

File: Hashtable/tasks/threeSum.java
/**
 * 三数之和 利用HashSet
 * 给定一个数组 nums = [-1, 0, 1, 2, -1, -4],

 * 一个解答方案是:
 * [
    [-1, 0, 1],
    [-1, -1, 2]
 * ] 
*/

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        // 定义一个集合 集合里面没有重复的元素
        Set<List<Integer>> res = new HashSet<>();
        // 没有3个数，直接返回空数组列表
        if (nums.length < 3) {
            return new ArrayList<>();
        }
        // 对数组排序
        Arrays.sort(nums);

        for (int i = 0; i < nums.length - 2; i++) {
            int j = i + 1; // 第二个位置开始
            int k = nums.length - 1; // 最后一个元素
            
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                
                if (sum == 0) {
                    // 使用后++ -- ,运算后再加或者减
                    res.add(Arrays.asList(nums[i], nums[j++], nums[k--]));
                } else if (sum > 0) { // 三数之和大于0，k要变小(nums[j] 已经是最小)
                    k--;
                } else if (sum < 0) { // 三数之和小于0，j变大
                    j++;
                }
            }
        }
        return new ArrayList<>(res);
    }
}